src/nix.rs

```diff
@@ -35,29 +35,32 @@
 use serde_json;
 use std::collections::HashMap;
 use std::ffi::OsStr;
-use std::io::BufRead;
+use std::os::unix::ffi::OsStrExt;
 use std::path::{Path, PathBuf};
-use std::process::{Command, Stdio};
+use std::process::Command;
 use vec1::Vec1;
 
+use StorePath;
+
 /// Execute Nix commands using a builder-pattern abstraction.
 #[derive(Clone)]
-pub struct CallOpts {
-    input: Input,
+pub struct CallOpts<'a> {
+    input: Input<'a>,
     attribute: Option<String>,
+    result_gc_root_path: Option<PathBuf>,
```

Comment:
Right off the bat feedback: this isn't workable. We need to GC root everything Nix evaluates and builds, and enforce it at the type system level.

One improvement on the way it works on master is if a store path result from the Nix evaluator carries with it its tempdir handle, forcing the tempdir to stay alive until the Nix value is garbage collected.

Reply:
So, what effectively happened is: Once I removed the `lorri shell` subcommand, no code whatsoever is using this stuff anymore. The only part that creates a gc root is the store path generated by `logged-evaluation.nix`, and that is handled by `roots`. I’m thinking about completely removing this from the builder for now.

Reply:
Okay, the problem with creating the roots after a build is that there is a small period of time (race condition) where a GC could delete the store path. Thus we must pass the gc root to the nix build.

Reply:
This is resolved by basing this branch on https://github.com/target/lorri/pull/131. PTAL